File: EstagioJaAPI/Models/Vaga.cs
using System.Text.Json.Serialization;
using EstagioJaAPI.Dtos;
using NHibernate.Mapping.Attributes;
using NHibernate.Util;

namespace EstagioJaAPI.Models;

[Class(Table = "tb_vagas")]
public class Vaga 
{

    public Vaga(){}

    public Vaga(string titulo, Turno turno, string descricao, IEnumerable<Competencia> requisitos, IEnumerable<Curso> cursos, string responsabilidades, string beneficios, string status, double valorDaBolsa, Modalidade modalidade)
    {
        this.titulo = titulo;
        this.descricao = descricao;
        this.requisitos = requisitos;
        this.cursos = cursos;
        this.responsabilidades = responsabilidades;
        this.beneficios = beneficios;
        this.status = status;
        this.valorDaBolsa = valorDaBolsa;
        this.modalidade = modalidade;
        this.turno = turno;
    }

    [JsonConstructorAttribute]
    public Vaga(int? id, string titulo, Turno turno, string descricao, IEnumerable<Competencia> requisitos, IEnumerable<Curso> cursos, 
        string responsabilidades, string beneficios, string status, double valorDaBolsa, Modalidade modalidade, IEnumerable<Estudante> estudantes,
        Empresa empresa)
    {
        this.id = id;
        this.titulo = titulo;
        this.descricao = descricao;
        this.requisitos = requisitos;
        this.cursos = cursos;
        this.responsabilidades = responsabilidades;
        this.beneficios = beneficios;
        this.status = status;
        this.valorDaBolsa = valorDaBolsa;
        this.modalidade = modalidade;
        this.turno = turno;
        this.estudantes = estudantes;
        this.empresa = empresa;
    }

    public Vaga(int? id, string titulo, Turno turno, string descricao, IEnumerable<Competencia> requisitos, IEnumerable<Curso> cursos,
        string responsabilidades, string beneficios, string status, double valorDaBolsa, Modalidade modalidade, DateTime prazo)
    {
        this.id = id;
        this.titulo = titulo;
        this.descricao = descricao;
        this.requisitos = requisitos;
        this.cursos = cursos;
        this.responsabilidades = responsabilidades;
        this.beneficios = beneficios;
        this.status = status;
        this.valorDaBolsa = valorDaBolsa;
        this.modalidade = modalidade;
        this.turno = turno;
        this.prazo = prazo;
    }

    [Id(Name = "id", Column = "id_vaga", TypeType = typeof(int))]
    [Generator(1, Class = "native")]
    public virtual int? id { get; set; }

    [Property(Column = "titulo_vaga")]
    public virtual string titulo { get; set; }

    [Property(Column = "desc_vaga")]
    public virtual string descricao { get; set; }

    [Bag(0, Table = "tb_competencias_vaga")]
    [Key(1, Column = "id_vaga")]
    [ManyToMany(2, Column = "id_competencia", ClassType = typeof (Competencia))]
    public virtual IEnumerable<Competencia> requisitos { get; set; }

    [Bag(0, Table = "tb_cursos_vaga")]
    [Key(1, Column = "id_vaga")]
    [ManyToMany(2, Column = "id_curso", ClassType = typeof (Curso))]
    public virtual IEnumerable<Curso> cursos { get; set; }

    [Property(Column = "responsabilidades_vaga", NotNull = true)]
    public virtual string responsabilidades { get; set; }

    [Property(Column = "turno_vaga", NotNull = true)]
    public virtual Turno turno { get; set; }

    [Property(Column = "beneficios_vaga", NotNull = true)]
    public virtual string beneficios { get; set; }

    [Property(Column = "status_vaga", NotNull = true)]
    public virtual string status { get; set; }

    [Property(Column = "bolsa_vaga", NotNull = true)]
    public virtual double valorDaBolsa { get; set; }

    [Property(Column = "prazo_vaga", NotNull = true)]
    public virtual DateTime prazo { get; set; }

    [Property(Column = "modalidade_vaga")]
    public virtual Modalidade modalidade {get; set;}

    [Bag(0, Table = "tb_vagas_estudante", Inverse = true)]
    [Key(1, Column = "id_vaga")]
    [ManyToMany(2, Column = "id_estudante", ClassType = typeof (Estudante))]
    public virtual IEnumerable<Estudante> estudantes { get; set; }

    [ManyToOne(Column = "id_empresa_vaga")]
    public virtual Empresa empresa { get; set; }

    public static Vaga FromVagaDto(VagaDto dto)
    {
        Empresa empresa = new Empresa();
        empresa.id = dto.idEmpresa;
        return new Vaga(
            dto.id,
            dto.titulo,
            dto.turno,
            dto.descricao,
            Competencia.FromCompetenciaDtoList(dto.requisitos),
            Curso.FromCursoDtoList(dto.cursos),
            dto.responsabilidades,
            dto.beneficios,
            dto.status,
            dto.valorDaBolsa,
            dto.modalidade,
            dto.prazo
        );
    }

    public static VagaDto ToVagaDto(Vaga vaga)
    {
        return new VagaDto(
              vaga.id,
              vaga.titulo,
              vaga.turno,   
              vaga.descricao,
              Competencia.ToCompetenciaDtoList(vaga.requisitos),
              Curso.ToCursoDtoList(vaga.cursos),
              vaga.responsabilidades,
              vaga.beneficios,
              vaga.status,
              vaga.valorDaBolsa,
              vaga.modalidade,
              vaga.empresa.id,
              vaga.prazo
        );
    } 

    public static Vaga FromVagaCandidaturaDto(VagaCandidaturaDto dto) {
        return new Vaga(
            dto.id,
            dto.titulo,
            dto.turno,
            dto.descricao,
            Competencia.FromCompetenciaDtoList(dto.requisitos),
            Curso.FromCursoDtoList(dto.cursos),
            dto.responsabilidades,
            dto.beneficios,
            dto.status,
            dto.valorDaBolsa,
            dto.modalidade,
            dto.prazo
        );
    } 

    public static VagaCandidaturaDto ToVagaCandidaturaDto(Vaga vaga)
    {
        return new VagaCandidaturaDto(
            vaga.id,
            vaga.titulo,
            vaga.turno,
            vaga.descricao,
            Competencia.ToCompetenciaDtoList(vaga.requisitos),
            Curso.ToCursoDtoList(vaga.cursos),
            vaga.responsabilidades,
            vaga.beneficios,
            vaga.status,
            vaga.valorDaBolsa,
            vaga.modalidade,
            vaga.empresa.id,
            vaga.empresa.nomeFantasia,
            Estudante.GetIdFromList(vaga.estudantes),
            vaga.prazo,
            vaga.estudantes.Count()
        );
    }

}

File: estagioja_flutter_app/lib/vaga/data/model/vaga.dart
import 'dart:convert';

import 'package:estagioja_flutter_app/shared/data/model/curso.dart';
import 'package:estagioja_flutter_app/shared/data/model/competencia.dart';
import 'package:estagioja_flutter_app/shared/data/model/modalidade.dart';
import 'package:estagioja_flutter_app/shared/data/model/turno.dart';

class Vaga {

  Vaga(this._id, this._titulo, this._descricao, this._requisitos, this._cursos,
      this._responsabilidades, this._beneficios, this._status, this._valorDaBolsa,
      this._modalidade, this._turno, this._idEmpresa, this._prazo, this._quantidadeCandidaturas,
      this._nomeEmpresa);

  int _id;
  String _titulo;
  String _descricao;
  List<Competencia> _requisitos;
  List<Curso> _cursos;
  String _responsabilidades;
  String _beneficios;
  String _status;
  double _valorDaBolsa;
  Modalidade _modalidade;
  Turno _turno;
  int _idEmpresa;
  DateTime _prazo;
  int _quantidadeCandidaturas;
  String _nomeEmpresa;

  Map<String, dynamic> toJson() => {
    'id': _id,
    'titulo': _titulo,
    'descricao': _descricao,
    'requisitos': jsonEncode(_requisitos.map((requisito) => requisito.toJson()).toList()),
    'cursos': jsonEncode(_cursos.map((curso) => curso.toJson()).toList()),
    'responsabilidade': _responsabilidades,
    'beneficios': _beneficios,
    'status': _status,
    'valorDaBolsa': _valorDaBolsa,
    'modalidade': _modalidade,
    'turno': _turno,
    'idEmpresa': _idEmpresa,
    'prazo': _prazo,
    'quantidadeCandidaturas': _quantidadeCandidaturas,
    'nomeEmpresa': _nomeEmpresa
  };

  Vaga.fromJson(Map<String, dynamic> json) :
    _id = json['id'],
    _titulo = json['titulo'],
    _descricao = json['descricao'],
    _requisitos = List<Competencia>.from((json['requisitos']).map((model) => Competencia.fromJson(model))),
    _cursos = List<Curso>.from((json['cursos']).map((model) => Curso.fromJson(model))),
    _responsabilidades = json['responsabilidades'],
    _beneficios = json['beneficios'],
    _status = json['status'],
    _valorDaBolsa = double.parse(json['valorDaBolsa'].toString()),
    _modalidade = Modalidade.values[json['modalidade']],
    _turno = Turno.values[json['turno']],
    _idEmpresa = json['idEmpresa'],
    _prazo = DateTime.parse(json['prazo']),
    _quantidadeCandidaturas = json['quantidadeCandidaturas'],
    _nomeEmpresa = json['nomeEmpresa']
  ;

  int get id => _id;

  set id(int value) {
    _id = value;
  }

  String get titulo => _titulo;

  set titulo(String value) {
    _titulo = value;
  }

  String get descricao => _descricao;

  set descricao(String value) {
    _descricao = value;
  }

  List<Competencia> get requisitos => _requisitos;

  set requisitos(List<Competencia> value) {
    _requisitos = value;
  }

  List<Curso> get cursos => _cursos;

  set cursos(List<Curso> value) {
    _cursos = value;
  }

  String get responsabilidades => _responsabilidades;

  set responsabilidades(String value) {
    _responsabilidades = value;
  }

  String get beneficios => _beneficios;

  set beneficios(String value) {
    _beneficios = value;
  }

  String get status => _status;

  set status(String value) {
    _status = value;
  }

  double get valorDaBolsa => _valorDaBolsa;

  set valorDaBolsa(double value) {
    _valorDaBolsa = value;
  }

  Modalidade get modalidade => _modalidade;

  set modalidade(Modalidade value) {
    _modalidade = value;
  }

  Turno get turno => _turno;

  set turno(Turno value) {
    _turno = value;
  }

  int get idEmpresa => _idEmpresa;

  set idEmpresa(int value) {
    _idEmpresa = value;
  }

  DateTime get prazo => _prazo;

  set prazo(DateTime value) {
    _prazo = value;
  }

  int get quantidadeCandidaturas => _quantidadeCandidaturas;

  set quantidadeCandidaturas(int value) {
    _quantidadeCandidaturas = value;
  }

  String get nomeEmpresa => _nomeEmpresa;

  set nomeEmpresa(String value) {
    _nomeEmpresa = value;
  }
}